File: OpenBuildSystemJava/src/org/abstractica/openbuildsystem/library/parts/sourced/fasteners/bolts/hexbolts/HexBolt.java
package org.abstractica.openbuildsystem.library.parts.sourced.fasteners.bolts.hexbolts;

import org.abstractica.openbuildsystem.core.Core;
import org.abstractica.openbuildsystem.core.Geometry3D;
import org.abstractica.openbuildsystem.core.Node3D;
import org.abstractica.openbuildsystem.library.parts.ACutoutPart;
import org.abstractica.openbuildsystem.library.parts.sourced.fasteners.bolts.Bolt;
import org.abstractica.openbuildsystem.library.util.math.HexMath;

public class HexBolt extends ACutoutPart implements Bolt
{
	private final double diameter, length, threadLength, headWidth, headHeight;

	public HexBolt(String sizeName,
	               double diameter,
	               double length,
	               double threadLength,
	               double headWidth,
	               double headHeight)
	{
		super("HexBolt_" + sizeName);
		this.diameter = diameter;
		this.length = length;
		this.threadLength = threadLength;
		this.headWidth = headWidth;
		this.headHeight = headHeight;
	}

	@Override
	public double getDiameter()
	{
		return diameter;
	}

	@Override
	public double getLength()
	{
		return length;
	}

	@Override
	public double getThreadLength()
	{
		return threadLength;
	}

	@Override
	public double getHeadDiameter()
	{
		return HexMath.hexagon_Diameter(headWidth);
	}

	@Override
	public double getHeadHeight()
	{
		return headHeight;
	}

	@Override
	protected Geometry3D doGenerateGeometry()
	{
		Node3D res = Core.union3D();
		//Head
		Node3D headColor = Core.color("silver");
		Node3D headTranslate = Core.translate3D(0, 0, -headHeight);
		Geometry3D head = Core.cylinder3D(HexMath.hexagon_Diameter(headWidth), headHeight, 6);
		headColor.add(head);
		headTranslate.add(headColor);
		res.add(headTranslate);

		//Non threaded part
		if (length > threadLength)
		{
			Geometry3D nonThreaded = Core.cylinder3D(diameter, length - threadLength, 32);
			Node3D nonThreadedColor = Core.color("silver");
			nonThreadedColor.add(nonThreaded);
			res.add(nonThreadedColor);
		}
		//Threaded part
		Node3D threadedTranslate = Core.translate3D(0, 0, length - threadLength);
		Geometry3D threaded = Core.cylinder3D(diameter, threadLength, 32);
		Node3D threadedColor = Core.color("grey");
		threadedColor.add(threaded);
		threadedTranslate.add(threadedColor);
		res.add(threadedTranslate);

		return res;
	}

	@Override
	protected Geometry3D doGenerateCutout(double fittingXY, double fittingZ, boolean fixOverhang)
	{
		Node3D res = Core.union3D();
		//Head
		Node3D headTranslate = Core.translate3D(0, 0, -headHeight - fittingZ);
		Geometry3D head = Core.cylinder3D(HexMath.hexagon_Diameter(headWidth + 2 * fittingXY), headHeight + 2 * fittingZ, 6);
		headTranslate.add(head);

		//Body
		Geometry3D body = Core.cylinder3D(diameter + 2 * fittingXY, length + fittingZ, 64);
		res.add(body);

		//fix overhang
		if (fixOverhang)
		{
			Node3D hull = Core.hull3D();
			hull.add(headTranslate);
			double height = 0.5 * (HexMath.hexagon_Diameter(headWidth - diameter));
			Geometry3D cyl = Core.cylinder3D(diameter + 2 * fittingXY, height, 64);
			hull.add(cyl);
			res.add(hull);
		} else
		{
			res.add(headTranslate);
		}
		return res;
	}
}

File: Kotlin/src/main/kotlin/obs/HexBolt.kt
package obs

import CodeBuilder
import core.Builder3D
import core.Component3D
import core.Core

class HexBolt(val diameter: Double,
              val length: Double,
              val thread_length: Double,
              val head_width: Double,
              val head_height: Double) : Builder3D<Core>
{
    override fun build(lib: Core): Component3D
    {
        val res = lib.union()
        //Non-threaded part
        if (thread_length < length)
        {
            var non_thread = lib.cylinder(diameter, length - thread_length, 32)
            non_thread = lib.color("silver").add(non_thread)
            res.add(non_thread)
        }
        //Threaded part
        var thread =  lib.cylinder(diameter, thread_length, 32)
        thread = lib.color("grey").add(thread)
        thread = lib.translate(0.0, 0.0, length - thread_length).add(thread)
        //Head
        var head = lib.cylinder(lib.hex_width_to_diameter(head_width), head_height, 6)
        head = lib.color("silver").add(head)
        head =  lib.translate(0.0, 0.0, -head_height).add(head)
        res.add(thread)
        res.add(head)
        return res
    }
/*
    override fun build(lib: Core): Component3D
    {
        val res = lib.union()
        //Non-threaded part
        if (thread_length < length)
        {
            res.add(
                    lib.color("silver").add(
                            lib.cylinder(diameter, length - thread_length, 32)
                    )
            )
        }
        //Threaded part
        res.add(
                lib.translate(0.0, 0.0, length - thread_length).add(
                        lib.color("grey").add(
                                lib.cylinder(diameter, thread_length, 32)
                        )
                )
        )
        //Head
        res.add(
                lib.translate(0.0, 0.0, -head_height).add(
                        lib.color("silver").add(
                                lib.cylinder(lib.hex_width_to_diameter(head_width), head_height, 6)
                        )
                )
        )
        return res
    }
    */
}